File: simulator/trahens_crypto/advertisement.py
# SPDX-License-Identifier: Apache-2.0
"""Trahens B1.2 discovery advertisement reference.

`network-bootstrap-b1.md` section 6 fixes what an advertisement may carry and
what it must not. ADR 0045 D4 makes it a fixed-width datagram padded to the
cell length, and D5 has it advertise under a short-lived key rather than a
long-term identity.

The datagram is one cell wide because discovery precedes any link, so there is
no encryption to hide a length under. A variable-length advertisement would
leak its shape on the wire and would make captures non-uniform, which the
harness's 1,052-byte assertion relies on.

**What is deliberately absent.** The advertisement carries no binding from the
short-lived key to the admission identity it will eventually use. That binding
is D5's "signed transition", and it cannot live here: putting the admission
static key in an unauthenticated datagram is exactly the stable network-wide
identifier section 6 forbids, and exactly what advertising under the long-term
key was rejected for. The transition belongs inside the handshake transcript,
where it is protected. This module therefore proves only that the advertiser
holds the short-lived key and that the fields have not been altered.

This reference exists to publish vectors. It is not independently audited and
MUST NOT be used as production security code.
"""

from __future__ import annotations

from dataclasses import dataclass

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

from trahens_spec.generated import (
    B12_DATAGRAM_ADVERTISEMENT,
    BYTES_B12_ADVERTISEMENT,
    BYTES_B12_ADVERTISEMENT_BODY,
    BYTES_B12_ADVERTISEMENT_SIGNATURE,
    BYTES_B12_COOKIE,
    DOMAIN_B12_ADVERTISEMENT,
)

MAX_LIST = 8


class AdvertisementError(ValueError):
    """Any failure. A reader never says which check refused a datagram."""


@dataclass(frozen=True)
class Advertisement:
    """What section 6 permits, and nothing else.

    No descriptor, no capability, no route label, and no stable identifier:
    `key` is short-lived and is the only identity here.
    """

    version: int
    key: bytes
    expiry_ms: int
    capacity_class: int
    auth_modes: int
    w2_profiles: tuple[int, ...]
    t1_profiles: tuple[int, ...]
    t2_profiles: tuple[int, ...]
    suites: tuple[int, ...]
    cookie: bytes | None = None


def _list_bytes(values: tuple[int, ...], width: int) -> bytes:
    if not 1 <= len(values) <= MAX_LIST:
        raise AdvertisementError("profile list out of range")
    out = bytes([len(values)])
    for value in values:
        if not 0 <= value < 256**width:
            raise AdvertisementError("profile identifier out of range")
        out += value.to_bytes(width, "big")
    return out


def _take_list(data: bytes, cursor: int, width: int) -> tuple[tuple[int, ...], int]:
    if cursor >= len(data):
        raise AdvertisementError("truncated profile list")
    count = data[cursor]
    cursor += 1
    if not 1 <= count <= MAX_LIST:
        raise AdvertisementError("profile list out of range")
    values = []
    for _ in range(count):
        chunk = data[cursor : cursor + width]
        if len(chunk) != width:
            raise AdvertisementError("truncated profile list")
        values.append(int.from_bytes(chunk, "big"))
        cursor += width
    return tuple(values), cursor


def encode_body(advertisement: Advertisement) -> bytes:
    if not 0 <= advertisement.version < 256:
        raise AdvertisementError("version out of range")
    if len(advertisement.key) != 32:
        raise AdvertisementError("advertisement key must be 32 bytes")
    if not 0 <= advertisement.expiry_ms < 2**64:
        raise AdvertisementError("expiry out of range")
    if not 0 <= advertisement.capacity_class < 256:
        raise AdvertisementError("capacity class out of range")
    if not 0 <= advertisement.auth_modes < 256:
        raise AdvertisementError("auth modes out of range")
    body = (
        bytes([advertisement.version])
        + advertisement.key
        + advertisement.expiry_ms.to_bytes(8, "big")
        + bytes([advertisement.capacity_class, advertisement.auth_modes])
        + _list_bytes(advertisement.w2_profiles, 1)
        + _list_bytes(advertisement.t1_profiles, 1)
        + _list_bytes(advertisement.t2_profiles, 1)
        + _list_bytes(advertisement.suites, 2)
    )
    if advertisement.cookie is None:
        return body + bytes([0])
    if len(advertisement.cookie) != BYTES_B12_COOKIE:
        raise AdvertisementError("cookie must be the registry width")
    return body + bytes([1]) + advertisement.cookie


def decode_body(body: bytes) -> Advertisement:
    if len(body) < 44:
        raise AdvertisementError("truncated advertisement")
    version = body[0]
    key = body[1:33]
    expiry_ms = int.from_bytes(body[33:41], "big")
    capacity_class = body[41]
    auth_modes = body[42]
    cursor = 43
    w2, cursor = _take_list(body, cursor, 1)
    t1, cursor = _take_list(body, cursor, 1)
    t2, cursor = _take_list(body, cursor, 1)
    suites, cursor = _take_list(body, cursor, 2)
    if cursor >= len(body):
        raise AdvertisementError("truncated advertisement")
    present = body[cursor]
    cursor += 1
    if present == 0:
        cookie = None
    elif present == 1:
        cookie = body[cursor : cursor + BYTES_B12_COOKIE]
        if len(cookie) != BYTES_B12_COOKIE:
            raise AdvertisementError("truncated cookie")
        cursor += BYTES_B12_COOKIE
    else:
        raise AdvertisementError("cookie flag is not canonical")
    if cursor != len(body):
        raise AdvertisementError("trailing bytes after the advertisement")
    return Advertisement(
        version, key, expiry_ms, capacity_class, auth_modes, w2, t1, t2, suites, cookie
    )


def _frame(body: bytes) -> bytes:
    """Length, body, then zeros to the fixed width, as B1.1 frames a payload."""
    if len(body) + 2 > BYTES_B12_ADVERTISEMENT_BODY:
        raise AdvertisementError("advertisement does not fit its datagram")
    return (
        len(body).to_bytes(2, "big")
        + body
        + bytes(BYTES_B12_ADVERTISEMENT_BODY - 2 - len(body))
    )


def _unframe(framed: bytes) -> bytes:
    if len(framed) != BYTES_B12_ADVERTISEMENT_BODY:
        raise AdvertisementError("framed region is the wrong width")
    length = int.from_bytes(framed[:2], "big")
    if 2 + length > BYTES_B12_ADVERTISEMENT_BODY:
        raise AdvertisementError("declared length overruns the datagram")
    if framed[2 + length :] != bytes(BYTES_B12_ADVERTISEMENT_BODY - 2 - length):
        raise AdvertisementError("padding is not zero")
    return framed[2 : 2 + length]


def encode(advertisement: Advertisement, signing_seed: bytes) -> bytes:
    """One signed datagram, exactly `b12_advertisement` bytes.

    The signature covers the discriminator and the whole framed region, padding
    included, so neither the type byte nor the padding can be altered without
    detection.
    """
    if len(signing_seed) != 32:
        raise AdvertisementError("signing seed must be 32 bytes")
    framed = _frame(encode_body(advertisement))
    signed = bytes([B12_DATAGRAM_ADVERTISEMENT]) + framed
    signature = Ed25519PrivateKey.from_private_bytes(signing_seed).sign(
        DOMAIN_B12_ADVERTISEMENT + signed
    )
    datagram = signed + signature
    if len(datagram) != BYTES_B12_ADVERTISEMENT:
        raise AdvertisementError("datagram width mismatch")
    return datagram


def decode(datagram: bytes) -> Advertisement:
    """Parse and verify. The key that signed is the key the datagram carries.

    That is all this establishes: the advertiser holds the short-lived key and
    the fields are intact. It says nothing about which identity the advertiser
    will admit under, which is the transition the handshake carries.
    """
    if len(datagram) != BYTES_B12_ADVERTISEMENT:
        raise AdvertisementError("advertisement is the wrong width")
    if datagram[0] != B12_DATAGRAM_ADVERTISEMENT:
        raise AdvertisementError("not an advertisement")
    signed = datagram[: BYTES_B12_ADVERTISEMENT - BYTES_B12_ADVERTISEMENT_SIGNATURE]
    signature = datagram[BYTES_B12_ADVERTISEMENT - BYTES_B12_ADVERTISEMENT_SIGNATURE :]
    advertisement = decode_body(_unframe(signed[1:]))
    try:
        Ed25519PublicKey.from_public_bytes(advertisement.key).verify(
            signature, DOMAIN_B12_ADVERTISEMENT + signed
        )
    except (InvalidSignature, ValueError) as error:
        raise AdvertisementError("advertisement signature failed") from error
    return advertisement

File: implementation/rust/crates/admission-b12/src/advertisement.rs
// SPDX-License-Identifier: Apache-2.0
//! The discovery advertisement datagram.
//!
//! Implements `spec/discovery-advertisement-b12.md`. One cell wide, because
//! discovery precedes any link and there is no encryption to hide a length
//! under; the first byte is a discriminator from the range
//! `link-handshake-b1.md` section 3 reserves.
//!
//! Verifying one shows that the advertiser holds the short-lived key and that
//! the fields are intact. It carries no binding to the admission identity the
//! advertiser will later use, so a verified advertisement is a hint about where
//! to try rather than a statement about who will answer.

use crate::CookieError;
use protocol_registry::{
    B12_DATAGRAM_ADVERTISEMENT, BYTES_B12_ADVERTISEMENT, BYTES_B12_ADVERTISEMENT_BODY,
    BYTES_B12_ADVERTISEMENT_SIGNATURE, BYTES_B12_COOKIE, DOMAIN_B12_ADVERTISEMENT,
    LIMIT_MAX_OFFERED_PROFILES_PER_CLASS,
};
use trahens_crypto::{sign, verify, SecretBytes};

type Result<T> = std::result::Result<T, CookieError>;

/// What `network-bootstrap-b1.md` section 6 permits, and nothing else.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Advertisement {
    pub version: u8,
    /// The short-lived advertisement key of ADR 0045 D5, and the only identity
    /// this datagram carries.
    pub key: [u8; 32],
    pub expiry_ms: u64,
    pub capacity_class: u8,
    pub auth_modes: u8,
    pub w2_profiles: Vec<u8>,
    pub t1_profiles: Vec<u8>,
    pub t2_profiles: Vec<u8>,
    pub suites: Vec<u16>,
    pub cookie: Option<[u8; BYTES_B12_COOKIE]>,
}

fn push_list_u8(out: &mut Vec<u8>, values: &[u8]) -> Result<()> {
    if values.is_empty() || values.len() > LIMIT_MAX_OFFERED_PROFILES_PER_CLASS {
        return Err(CookieError);
    }
    out.push(u8::try_from(values.len()).map_err(|_| CookieError)?);
    out.extend_from_slice(values);
    Ok(())
}

fn push_list_u16(out: &mut Vec<u8>, values: &[u16]) -> Result<()> {
    if values.is_empty() || values.len() > LIMIT_MAX_OFFERED_PROFILES_PER_CLASS {
        return Err(CookieError);
    }
    out.push(u8::try_from(values.len()).map_err(|_| CookieError)?);
    for value in values {
        out.extend_from_slice(&value.to_be_bytes());
    }
    Ok(())
}

fn take_list_u8(body: &[u8], cursor: &mut usize) -> Result<Vec<u8>> {
    let count = usize::from(*body.get(*cursor).ok_or(CookieError)?);
    *cursor += 1;
    if count == 0 || count > LIMIT_MAX_OFFERED_PROFILES_PER_CLASS {
        return Err(CookieError);
    }
    let taken = body
        .get(*cursor..*cursor + count)
        .ok_or(CookieError)?
        .to_vec();
    *cursor += count;
    Ok(taken)
}

fn take_list_u16(body: &[u8], cursor: &mut usize) -> Result<Vec<u16>> {
    let count = usize::from(*body.get(*cursor).ok_or(CookieError)?);
    *cursor += 1;
    if count == 0 || count > LIMIT_MAX_OFFERED_PROFILES_PER_CLASS {
        return Err(CookieError);
    }
    let mut values = Vec::with_capacity(count);
    for _ in 0..count {
        let pair = body.get(*cursor..*cursor + 2).ok_or(CookieError)?;
        values.push(u16::from_be_bytes([pair[0], pair[1]]));
        *cursor += 2;
    }
    Ok(values)
}

impl Advertisement {
    fn encode_body(&self) -> Result<Vec<u8>> {
        let mut body = Vec::with_capacity(64);
        body.push(self.version);
        body.extend_from_slice(&self.key);
        body.extend_from_slice(&self.expiry_ms.to_be_bytes());
        body.push(self.capacity_class);
        body.push(self.auth_modes);
        push_list_u8(&mut body, &self.w2_profiles)?;
        push_list_u8(&mut body, &self.t1_profiles)?;
        push_list_u8(&mut body, &self.t2_profiles)?;
        push_list_u16(&mut body, &self.suites)?;
        match self.cookie {
            None => body.push(0),
            Some(cookie) => {
                body.push(1);
                body.extend_from_slice(&cookie);
            }
        }
        Ok(body)
    }

    fn decode_body(body: &[u8]) -> Result<Self> {
        if body.len() < 44 {
            return Err(CookieError);
        }
        let key: [u8; 32] = body
            .get(1..33)
            .ok_or(CookieError)?
            .try_into()
            .map_err(|_| CookieError)?;
        let expiry_bytes: [u8; 8] = body
            .get(33..41)
            .ok_or(CookieError)?
            .try_into()
            .map_err(|_| CookieError)?;
        let mut cursor = 43_usize;
        let w2_profiles = take_list_u8(body, &mut cursor)?;
        let t1_profiles = take_list_u8(body, &mut cursor)?;
        let t2_profiles = take_list_u8(body, &mut cursor)?;
        let suites = take_list_u16(body, &mut cursor)?;
        let present = *body.get(cursor).ok_or(CookieError)?;
        cursor += 1;
        let cookie = match present {
            0 => None,
            1 => {
                let taken: [u8; BYTES_B12_COOKIE] = body
                    .get(cursor..cursor + BYTES_B12_COOKIE)
                    .ok_or(CookieError)?
                    .try_into()
                    .map_err(|_| CookieError)?;
                cursor += BYTES_B12_COOKIE;
                Some(taken)
            }
            // Any other flag value is refused rather than read as present.
            _ => return Err(CookieError),
        };
        if cursor != body.len() {
            return Err(CookieError);
        }
        Ok(Self {
            version: body[0],
            key,
            expiry_ms: u64::from_be_bytes(expiry_bytes),
            capacity_class: body[41],
            auth_modes: body[42],
            w2_profiles,
            t1_profiles,
            t2_profiles,
            suites,
            cookie,
        })
    }
}

/// One signed datagram, exactly `b12_advertisement` bytes.
///
/// The signature covers the discriminator and the whole framed region, padding
/// included, so neither the type byte nor the padding can be altered without
/// detection.
pub fn encode(advertisement: &Advertisement, signing_secret: &SecretBytes<64>) -> Result<Vec<u8>> {
    let body = advertisement.encode_body()?;
    if body.len() + 2 > BYTES_B12_ADVERTISEMENT_BODY {
        return Err(CookieError);
    }
    let mut datagram = Vec::with_capacity(BYTES_B12_ADVERTISEMENT);
    datagram.push(B12_DATAGRAM_ADVERTISEMENT);
    datagram.extend_from_slice(
        &u16::try_from(body.len())
            .map_err(|_| CookieError)?
            .to_be_bytes(),
    );
    datagram.extend_from_slice(&body);
    datagram.resize(
        BYTES_B12_ADVERTISEMENT - BYTES_B12_ADVERTISEMENT_SIGNATURE,
        0,
    );

    let mut signed = Vec::with_capacity(DOMAIN_B12_ADVERTISEMENT.len() + datagram.len());
    signed.extend_from_slice(DOMAIN_B12_ADVERTISEMENT);
    signed.extend_from_slice(&datagram);
    datagram.extend_from_slice(&sign(signing_secret, &signed)?);
    if datagram.len() != BYTES_B12_ADVERTISEMENT {
        return Err(CookieError);
    }
    Ok(datagram)
}

/// Parse and verify. The key that signed is the key the datagram carries.
pub fn decode(datagram: &[u8]) -> Result<Advertisement> {
    if datagram.len() != BYTES_B12_ADVERTISEMENT
        || datagram.first() != Some(&B12_DATAGRAM_ADVERTISEMENT)
    {
        return Err(CookieError);
    }
    let split = BYTES_B12_ADVERTISEMENT - BYTES_B12_ADVERTISEMENT_SIGNATURE;
    let head = datagram.get(..split).ok_or(CookieError)?;
    let signature: [u8; BYTES_B12_ADVERTISEMENT_SIGNATURE] = datagram
        .get(split..)
        .ok_or(CookieError)?
        .try_into()
        .map_err(|_| CookieError)?;

    let framed = head.get(1..).ok_or(CookieError)?;
    let length = usize::from(u16::from_be_bytes([
        *framed.first().ok_or(CookieError)?,
        *framed.get(1).ok_or(CookieError)?,
    ]));
    if 2 + length > framed.len() {
        return Err(CookieError);
    }
    // The padding is inside the signed region, but it is checked here as well:
    // a receiver must not accept a datagram whose declared length leaves
    // anything but zeros behind it.
    if framed
        .get(2 + length..)
        .ok_or(CookieError)?
        .iter()
        .any(|byte| *byte != 0)
    {
        return Err(CookieError);
    }
    let advertisement = Advertisement::decode_body(framed.get(2..2 + length).ok_or(CookieError)?)?;

    let mut signed = Vec::with_capacity(DOMAIN_B12_ADVERTISEMENT.len() + head.len());
    signed.extend_from_slice(DOMAIN_B12_ADVERTISEMENT);
    signed.extend_from_slice(head);
    verify(&advertisement.key, &signed, &signature)?;
    Ok(advertisement)
}
